Donor/Repositories/DonorRepository.cs

```diff
@@ -1,80 +1,105 @@
 {
     public class DonorRepository : IDonorRepository
     {
         private readonly DonorContext _context;
-		private readonly ILogger<DonorRepository> _log;
+        private readonly ILogger<DonorRepository> _log;
 
-		public DonorRepository(DonorContext context, ILogger<DonorRepository> log)
+        public DonorRepository(DonorContext context, ILogger<DonorRepository> log)
         {
             _context = context ?? throw new ArgumentNullException(nameof(context));
-			_log = log ?? throw new ArgumentNullException(nameof(log));
-		}
-	
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+        }
 
-        public async Task AddDonorAsync(Entities.Donor donor)
+        public async Task AddDonorAsync(Entities.Donor donor, List<int> donationPreferences)
         {
-			try
-			{
-				await _context.Donors.AddAsync(donor);				
+            using var transaction = await _context.Database.BeginTransactionAsync();
+            try
+            {
+                await _context.Donors.AddAsync(donor);
+                await _context.SaveChangesAsync();
+
+                if (donationPreferences != null && donationPreferences.Any())
+                {
+                    foreach (var organId in donationPreferences)
+                    {
+                        var preference = new DonationPreference
+                        {
+                            DonorId = donor.Id,
+                            OrganId = organId
+                        };
+                        await _context.DonationPreferences.AddAsync(preference);
+                    }
+                    await _context.SaveChangesAsync();
+                }
 
-			}			
-			catch (Exception ex)
-			{
-				_log.LogError(ex, "Error registering donor");
-				throw;
-			}
-			
-			
+                await transaction.CommitAsync();
+            }
+            catch (DbUpdateException ex) when (IsUniqueConstraintViolation(ex))
+            {
+                _log.LogError(ex, "Database update error while adding donor");
+                await transaction.RollbackAsync();
+                throw new ValidationException("A donor with the same Identity Number or Email already exists.");
+            }
+            catch (Exception ex)
+            {
+                _log.LogError(ex, "Error adding donor");
+                await transaction.RollbackAsync();
+                throw;
+            }
         }
 
         public async Task<IEnumerable<Entities.Donor>> GetAllDonorsAsync()
         {
-			IEnumerable<Entities.Donor > donors = new List<Entities.Donor>();
-			try
-			{
-				donors = await _context.Donors.ToListAsync();
-			}
-			catch (Exception ex)
-			{
-
-				_log.LogError(ex, "Error getting all donors");
-				throw;
-			}
+            IEnumerable<Entities.Donor> donors = new List<Donor.Entities.Donor>();
+            try
+            {
+                donors = await _context.Donors
+                    .Include(d => d.DonationPreferences)
```

Comment:
.Include(d => d.DonationPreferences) will not be necessary when you make the correction on the relationship between Donor and Organ